src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/DirectiveSyntaxExtensions.cs

```diff
@@ -2,82 +2,179 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
 using System.Runtime.CompilerServices;
 using System.Threading;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 
 namespace Microsoft.CodeAnalysis.CSharp.Extensions
 {
     internal static partial class DirectiveSyntaxExtensions
     {
-        private static readonly ConditionalWeakTable<SyntaxNode, DirectiveInfo> s_rootToDirectiveInfo =
-            new();
+        private static readonly ConditionalWeakTable<SyntaxNode, DirectiveInfo> s_rootToDirectiveInfo = new();
+        private static readonly ObjectPool<Stack<DirectiveTriviaSyntax>> s_stackPool = new(() => new());
 
         private static SyntaxNode GetAbsoluteRoot(this SyntaxNode node)
         {
-            while (node != null && (node.Parent != null || node is StructuredTriviaSyntax))
+            while (node.Parent != null || node is StructuredTriviaSyntax)
             {
-                if (node.Parent != null)
-                {
-                    node = node.Parent;
-                }
-                else
-                {
-                    node = node.ParentTrivia.Token.Parent;
-                }
+                node = node.Parent ?? node.ParentTrivia.Token.GetRequiredParent();
             }
 
             return node;
         }
 
         private static DirectiveInfo GetDirectiveInfo(SyntaxNode node, CancellationToken cancellationToken)
+            => s_rootToDirectiveInfo.GetValue(
+                node.GetAbsoluteRoot(),
+                root => GetDirectiveInfoForRoot(root, cancellationToken));
+
+        private static DirectiveInfo GetDirectiveInfoForRoot(SyntaxNode root, CancellationToken cancellationToken)
         {
-            var root = node.GetAbsoluteRoot();
-            var info = s_rootToDirectiveInfo.GetValue(root, r =>
+            var directiveMap = new Dictionary<DirectiveTriviaSyntax, DirectiveTriviaSyntax?>(
```

Comment:
this is all the logic from the walker, just moved into inline state in this method.